### PR TITLE
Fixed findbugs issues

### PR DESCRIPTION
**Pull request by rsandell**
Configured findbugs for xml output so Jenkins can pick them up.
And updated the maven release plugin.

@reviewbybees

### CODE REVIEW

**Comment by jtnord** [comment]
:bee:

File: pom.xml [hunk 1]
```diff
@@ -60,6 +65,26 @@
             <url>http://repo.jenkins-ci.org/public/</url>
         </pluginRepository>
     </pluginRepositories>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-release-plugin</artifactId>
+                <version>2.5.1</version>
```
Comment (stephenc) [inline pom.xml]: 2.5.2 now :-)

**Comment by stephenc** [comment]
:bee:

**Comment by rsandell** [comment]
@reviewbybees done

**Comment by ikedam** [comment]
I want to see how findbug works.
Could you let me know how to do that?

I could not get bug reports with `mvn findbugs:findbugs` even removing `SuppressFBWarnings`.

**Comment by ikedam** [comment]
@reviewbybees Please consider to make a guideline for your employees to create good pull requests.
Some additional information in pull requests helps developers much.

1. Contexts of requests.
    * It is often done with creating a JIRA issue.
    * In this case, it can be done with pointing https://groups.google.com/forum/#!searchin/jenkinsci-dev/findbugs/jenkinsci-dev/HxxTICM4LVQ/p7fNQhDDLZMJ.
    * And I suppose you are now applying findbugs changes to all repositories (maybe in the alphabetical order?)
2. What is findbugs? How does the change work?
    * Showing an example build failed with findbugs makes developers to know what findbugs is.
3. Why do you request the change?
    * "And updated the maven release plugin" can be easily seen from "Files changed" tabs. Developers rather want to know "Why you request that change". I suppose this is to avoid SNAPSHOT problem: https://wiki.jenkins-ci.org/display/JENKINS/Hosting+Plugins#HostingPlugins-Workingaroundcommonissues.

**Comment by ikedam** [comment]
I found that I had to generate class files first.
```
mvn package -Dmaven.test.skip=true
mvn findbugs:findbugs -Dfindbugs.jvmArgs="-Duser.language=en -Duser.country=US"
````
works well.
Please let me know if there's an easier way.

And I found that findbug fails with following errors on Windows, Japanese locale:
```
[Fatal Error] findbugsXml.xml:2:21349: Invalid byte 1 of 1-byte UTF-8 sequence.
```
findbugsXml.xml looks generated with ShiftJIS (a Japanese character encoding).
It'll annoy me if findbug works even for releasing.
I may have to define `outputEncoding` for findbugs.

File: src/main/java/hudson/plugins/build_timeout/impl/ElasticTimeOutStrategy.java [hunk 2]
```diff
@@ -95,7 +95,7 @@ private double averageDuration(AbstractBuild<?, ?> build, BuildListener listener
         }
 
 
-        return nonFailingBuilds > 0 ? durationSum / nonFailingBuilds : 0;
+        return nonFailingBuilds > 0 ? ((double)durationSum) / nonFailingBuilds : 0;
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/impl/ElasticTimeOutStrategy.java]: This is a fix also for findbugs to suppress the following error:
```
<BugInstance
  type='ICAST_IDIV_CAST_TO_DOUBLE'
  priority='Normal'
  category='STYLE'
  message='Integral division result cast to double or float in hudson.plugins.build_timeout.impl.ElasticTimeOutStrategy.averageDuration(AbstractBuild, BuildListener)'
  lineNumber='98'/>
```

File: src/main/java/hudson/plugins/build_timeout/BuildTimeoutWrapper.java [hunk 3]
```diff
@@ -41,7 +42,8 @@
  * @author Kohsuke Kawaguchi
  */
 public class BuildTimeoutWrapper extends BuildWrapper {
-    
+
+    @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Diagnostic fields are left mutable so that groovy console can be used to dynamically turn/off probes.")
     public static long MINIMUM_TIMEOUT_MILLISECONDS = Long.getLong(BuildTimeoutWrapper.class.getName()+ ".MINIMUM_TIMEOUT_MILLISECONDS", 3 * 60 * 1000);
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/BuildTimeoutWrapper.java]: This is a fix for the following findbugs error:
```
<BugInstance
  type='MS_SHOULD_BE_FINAL'
  priority='High'
  category='MALICIOUS_CODE'
  message='hudson.plugins.build_timeout.BuildTimeoutWrapper.MINIMUM_TIMEOUT_MILLISECONDS isn&apos;t final but should be'
```

This field is rather for testing purpose, to shorten testing duration.
We may consider to change this final and provide another way for testing.

**Comment by ikedam** [comment]
This looks good to me. I found fingbugs really useful.

**Comment by jtnord** [comment]
@ikedam the output encoding should default to project.reporting.outputEncoding.
It looks like this is not set in the parent and is falling back to the platform default - and hence why you are getting Japanese encoding.

As we can't change old releases of the parent the following may work.
```
  <properties>
    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
  </properties>
```
Note: - there presumably exists the same issue for the generated javadoc.

To view findbugs - you can do this in your IDE (eclipse, and intellij - I would guess that netbeans has something similar) or you can use `mvn findbugs:gui`

**Comment by ikedam** [comment]
@jtnord Thanks for the comment! I'll try that.

**Comment by ikedam** [comment]
Would someone enable findbugs feature on jenkins.ci.cloudbees.com for build-timeout-plugin?

I found that I had to ask that before merging this...I worked in a bad way, sorry.

**Comment by jtnord** [comment]
@ikedam enabled and set threshold to zero bugs.
[build 88](https://jenkins.ci.cloudbees.com/job/plugins/job/build-timeout-plugin/88/) has been kicked off to verify the changes.

**Comment by ikedam** [comment]
@jtnord Thanks a lot!
Some warnings look introduced in #40, and I'll work for that.

**Comment by jtnord** [comment]
@ikedam ok.  It also reports that the [source encoding](https://jenkins.ci.cloudbees.com/job/plugins/job/build-timeout-plugin/88/findbugsResult/error/) has not been set.
```
  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
  </properties>
```

**Comment by rsandell** [comment]
Sorry @ikedam , FindBugs is such a popular tool that I just assumed everyone knew about it and didn't need much explanation. I'll try to do better next time.
Those 7 new warnings that showed up in build 88 didn't when I ran it locally on my machine, I thought that I had fixed them all.

**Comment by ikedam** [comment]
findbugs-maven-plugin 3.0.1 looks contain a bug in encoding handling.
gleclaire/findbugs-maven-plugin#18
That reproduced also in my environment.

**Comment by rsandell** [comment]
Well its ```findbugs.xml``` that is the interesting one, ```findbugsXml.xml``` is IIUC of an older format that Jenkins can't use anyways.

**Pull request merged**